text/0005-firefox-data-stores.md

```diff
@@ -0,0 +1,21 @@
+
+# Problem
+
+There are a large number of data stores being used to hold data across all Firefox instances and platforms. There is currently nowhere that documents all these data stores and how they share information between them. We have 3 platforms, desktop, Android and iOS and each of these platforms shares data with the others. But not all the information is synced completely across all platforms, and the data itself is not stored in the same way everywhere. When starting to think about how we might provide a holistic user data solution for Mozilla, it is important to understand the scope of the problem and the specifics of the data that is being handled. 
+
+We therefore need to look at all of the data stores that we are currently using and document their purpose, structure, the kind of data that they hold and how they share that data.
+
+
+# Solution
+
+The extensive documentation can be found in the [Firefox Data Stores](https://github.com/mozilla/firefox-data-store-docs) repository.
+
+* There are a total of 46 separate data stores in Firefox for desktop. Within these 46 stores, there are 10 different storage formats. Very few of these stores are well documented in the code. 
+* Firefox for Android has only 3 stores as it consolidates many of the separate stores found in desktop Firefox into a single SQLite database. It uses only 3 different storage formats, SQLite, .json & .js. Android documents it's data stores well and uses a single file to define the schema for all it's tables.
+* Firefox for iOS has 5 data stores. One of these is historical as the Reading List store was reused from an earlier Mozilla app. One of the stores is used to store metadata for use in Activity Stream and has been created separately to make syncing of this data at a later data easier. It uses 2 different storage formats, SQLite and .json. iOS stores are not especially well documented, but the schemas are defined in a series of table or database specific files which makes discovering the schemas trivial. The exception to this is the reading list store which is defined separately from the others due to it history as part of another application.
+* Desktop stores logins in an encrypted json store, whereas Android has combined the logins table with it's main SQLite database. iOS keeps it as a separate SQLite database that uses sqlcipher for encryption. 
```

Comment:
`logins.json` is indeed migrated from `signons.sqlite` for desktop users bringing profiles forward from a pre-2014 version of Firefox, or in cases of bad dataloss (which, given the total lack of integration with Sync, is perhaps the worst thing we could do — it's a silent rollback, and attempts to fix it by futzing with Sync preferences will erase the new data on the server!).

On Android we have _two_ login stores currently in the tree. We have `PasswordsProvider`, which uses SQLite to share the Gecko-side `signins.sqlite`. And we have `LoginsProvider`, which is checked in and tested but not used. As mentioned in the commit message:

> but for now we just want to get the new provider landed and the tests running

I think we should primarily document the stores that are used, not the future migration path or the past migration paths. By all means mention the future, but it's not quite here yet.